### PR TITLE
pre_defined_classes-simple_programs_taking_string_user_input-medium-q6

### PR DESCRIPTION
**Pull request by yuqi88**
## PR Checklist
- [ ]  Are helper methods really needed here?
    > Opt for hardcoding input/output into the input provider when possible
- [ ]  Are the assertion fail messages good?
    - [ ]  Logical (does it make sense?)
    - [ ]  Readable (can a student understand what’s wrong?)
    - [ ]  Not too feed-y (does it give away the answer?)
    - [ ]  Grammatically sound (is the grammar and punctuation in the sentence proper?)
    - [ ]  Specific (does the fail message relate to the question?)
- [ ]  Is the code formatted well?
    > Run `CTRL` + `ALT` + `L` in IntelliJ
- [ ]  EOF newlines
- [ ]  Do the comments for Parsons follow the correct convention?
- [ ]  Is the question being tested sufficiently?
    - [ ]  Are there representative cases to test the program?
    - [ ]  What about edge cases? (empty strings, negatives, infinity, etc)
        > If the question doesn’t specify what to do, mark it as a bad question
---

## Not yet, but soon

- [ ]  Are the test method names good?
    > Test method names show up in a “What went well” section on the website, so “mathTest” will become “What went well: math test”, which doesn’t make much sense. A better name in this example would be “calculatesTaxCorrectly” so it shows up as “What went well: calculates tax correctly”

### CODE REVIEW

File: src/pre_defined_classes/simple_programs_taking_string_user_input/medium/q6/MainTest.java [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+    public void runMain() { HasSpace.main(new String[0]);}
+
+    static Stream<String> stringWithSpace() {
+        return Stream.of("The inner machinations of my mind are an enigma.", "(stu\\#:123 456)"
+                , "it\\'s a sunny day", "I think so\\.", " "
+        );
+    }
+
+    static Stream<String> stringWithNoSpace() {
+        return Stream.of("abc.", "1234", "EmptyStringDoesNotWork", "b");
```
Comment (Hedgemon4) [inline src/pre_defined_classes/simple_programs_taking_string_user_input/medium/q6/MainTest.java:38]: Should also test an empty string

File: src/pre_defined_classes/simple_programs_taking_string_user_input/medium/q6/MainTest.java [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+    }
+
+    @ParameterizedTest
+    @MethodSource("stringWithSpace")
+    public void correctlyDetectSpace(String sentence) throws InvalidClauseException {
+        TestOption.incorrectStructureErrorMessage = "Your program does not correctly print true when there is/are space(s) in the inputted sentence.";
```
Comment (Hedgemon4) [inline src/pre_defined_classes/simple_programs_taking_string_user_input/medium/q6/MainTest.java:44]: I would just say `... is a space in the inputted sentence.` as we only care if there is at least one.

File: src/pre_defined_classes/simple_programs_taking_string_user_input/medium/q6/MainTest.java [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+        runWithInput(sentence, new Clause[]{
+                new StringLiteral("true")
+        });
+    }
+
+    @ParameterizedTest
+    @MethodSource("stringWithNoSpace")
+    public void correctlyDetectNoSpace(String sentence) throws InvalidClauseException {
+        TestOption.incorrectStructureErrorMessage = "Your program does not correctly print false when there is/are space(s) in the inputted sentence.";
```
Comment (Hedgemon4) [inline src/pre_defined_classes/simple_programs_taking_string_user_input/medium/q6/MainTest.java:53]: Same as other error message

File: src/pre_defined_classes/simple_programs_taking_string_user_input/medium/q6/MainTest.java (hunk 1 above)
Comment (Hedgemon4) [reply src/pre_defined_classes/simple_programs_taking_string_user_input/medium/q6/MainTest.java:38]: I think as this is a nextLine() you will have to use System.lineSeperator()

File: src/pre_defined_classes/simple_programs_taking_string_user_input/medium/q6/HasSpace.java [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+    (stu#:123 456)
+    Does the sentence contain any spaces?
+    true
+Sample output:
+    Enter any sentence:
+    abc.
+    Does the sentence contain any spaces?
+    false
+
+   question: sometime boolean gets printed out as True instead of true? Should take that into consider...?
```
Comment (Hedgemon4) [inline src/pre_defined_classes/simple_programs_taking_string_user_input/medium/q6/HasSpace.java:23]: Not sure why this would be

**Review by Hedgemon4 — CHANGES_REQUESTED** [review]

File: src/pre_defined_classes/simple_programs_taking_string_user_input/medium/q6/HasSpace.java (hunk 4 above)
Comment (yuqi88) [reply src/pre_defined_classes/simple_programs_taking_string_user_input/medium/q6/HasSpace.java:23]: okay, there's no problem here. I mixed this up with Python.

Comment (Hedgemon4) [reply src/pre_defined_classes/simple_programs_taking_string_user_input/medium/q6/HasSpace.java]: Ok, all good

File: src/pre_defined_classes/simple_programs_taking_string_user_input/medium/q6/MainTest.java [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+    @ParameterizedTest
+    @MethodSource("stringWithNoSpace")
+    public void correctlyDetectNoSpace(String sentence) throws InvalidClauseException {
+        TestOption.incorrectStructureErrorMessage =
+                "Your program does not correctly print false when there is a space in the inputted sentence.";
+        runWithInput(sentence, new Clause[]{
+                new StringLiteral("false"),
+        });
+    }
+}
```
Comment (Hedgemon4) [inline src/pre_defined_classes/simple_programs_taking_string_user_input/medium/q6/MainTest.java:63]: Add an empty line to the end of the file

File: src/pre_defined_classes/simple_programs_taking_string_user_input/medium/q6/HasSpace.java [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+/*
+Languages that you and I speak are written in a way that uses whitespace to separate one word from one another.
+Make a program that answers true or false to the question of whether the inputted sentence has at least 1 space in it.
+
+Sample output:
+    Enter any sentence:
+    I think so.
+    Does the sentence contain any spaces?
+    true
+Sample output:
```
Comment (Hedgemon4) [inline src/pre_defined_classes/simple_programs_taking_string_user_input/medium/q6/HasSpace.java:12]: Only need one sample output

File: src/pre_defined_classes/simple_programs_taking_string_user_input/medium/q6/MainTest.java [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+        return Stream.of("abc.", "1234", "EmptyStringDoesNotWork", "b",
+                System.lineSeparator()
+        );
+    }
+
+    @ParameterizedTest
+    @MethodSource("stringWithSpace")
+    public void correctlyDetectSpace(String sentence) throws InvalidClauseException {
+        TestOption.incorrectStructureErrorMessage =
+                "Your program does not correctly print true when there is a space in the inputted sentence.";
```
Comment (Hedgemon4) [inline src/pre_defined_classes/simple_programs_taking_string_user_input/medium/q6/MainTest.java:48]: `Correctly print true` does not really make sense in this case

Comment (Hedgemon4) [reply src/pre_defined_classes/simple_programs_taking_string_user_input/medium/q6/MainTest.java:48]: I would change it to something similar to `Your program does not print the correct output when the string contains a space. `

Comment (Hedgemon4) [reply src/pre_defined_classes/simple_programs_taking_string_user_input/medium/q6/MainTest.java:48]: We want the message to relate to the question, but we also don't want to tell them the exact answer. They should have to go back and read the question to figure out what they need to change.

File: src/pre_defined_classes/simple_programs_taking_string_user_input/medium/q6/MainTest.java [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+        runWithInput(sentence, new Clause[]{
+                new StringLiteral("true")
+        });
+    }
+
+    @ParameterizedTest
+    @MethodSource("stringWithNoSpace")
+    public void correctlyDetectNoSpace(String sentence) throws InvalidClauseException {
+        TestOption.incorrectStructureErrorMessage =
+                "Your program does not correctly print false when there is a space in the inputted sentence.";
```
Comment (Hedgemon4) [inline src/pre_defined_classes/simple_programs_taking_string_user_input/medium/q6/MainTest.java:58]: Same as the other error message

File: src/pre_defined_classes/simple_programs_taking_string_user_input/medium/q6/MainTest.java [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+    static Stream<String> stringWithSpace() {
+        return Stream.of("The inner machinations of my mind are an enigma.", "(stu\\#:123 456)"
+                , "it\\'s a sunny day", "I think so\\.", " "
+        );
+    }
+
+    static Stream<String> stringWithNoSpace() {
+
+        return Stream.of("abc.", "1234", "EmptyStringDoesNotWork", "b",
+                System.lineSeparator()
```
Comment (Hedgemon4) [inline src/pre_defined_classes/simple_programs_taking_string_user_input/medium/q6/MainTest.java:40]: Did the `System.lineSeperator()` work?

**Review by Hedgemon4 — CHANGES_REQUESTED** [review]

File: src/pre_defined_classes/simple_programs_taking_string_user_input/medium/q6/MainTest.java (hunk 7 above)
Comment (yuqi88) [reply src/pre_defined_classes/simple_programs_taking_string_user_input/medium/q6/MainTest.java:48]: okie

File: src/pre_defined_classes/simple_programs_taking_string_user_input/medium/q6/MainTest.java (hunk 9 above)
Comment (yuqi88) [reply src/pre_defined_classes/simple_programs_taking_string_user_input/medium/q6/MainTest.java:40]: yep!

File: src/pre_defined_classes/simple_programs_taking_string_user_input/medium/q6/MainTest.java [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+        );
+    }
+
+    @ParameterizedTest
+    @MethodSource("stringWithSpace")
+    public void correctlyDetectSpace(String sentence) throws InvalidClauseException {
```
Comment (Hedgemon4) [inline src/pre_defined_classes/simple_programs_taking_string_user_input/medium/q6/MainTest.java:46]: Method name should be `correctlyDetectsSpace`

File: src/pre_defined_classes/simple_programs_taking_string_user_input/medium/q6/MainTest.java [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+        TestOption.incorrectStructureErrorMessage =
+                "Your program does not print the correct output when the string contains a space.";
+        runWithInput(sentence, new Clause[]{
+                new StringLiteral("true")
+        });
+    }
+
+    @ParameterizedTest
+    @MethodSource("stringWithNoSpace")
+    public void correctlyDetectNoSpace(String sentence) throws InvalidClauseException {
```
Comment (Hedgemon4) [inline src/pre_defined_classes/simple_programs_taking_string_user_input/medium/q6/MainTest.java:56]: `correctlyDetectsNoSpace`

**Review by Hedgemon4 — CHANGES_REQUESTED** [review]

**Review by Hedgemon4 — APPROVED** [review]

**Pull request merged**